File: baboon-compiler/src/main/resources/baboon-runtime/scala/BaboonExceptions.scala
package baboon.runtime.shared {

  trait BaboonCodecException extends Exception {
    def message: String
    override def getMessage: String = message
  }

  object BaboonCodecException {

    final case class EncoderFailure(message: String, cause: Throwable = null) extends Exception(message, cause) with BaboonCodecException

    final case class DecoderFailure(message: String, cause: Throwable = null) extends Exception(message, cause) with BaboonCodecException

    final case class ConverterFailure(message: String, cause: Throwable = null) extends Exception(message, cause) with BaboonCodecException

    final case class CodecNotFound(message: String) extends Exception(message) with BaboonCodecException

    final case class ConversionNotFound(message: String) extends Exception(message) with BaboonCodecException
  }
}

File: baboon-compiler/src/main/resources/baboon-runtime/kotlin/BaboonExceptions.kt
package baboon.runtime.shared

sealed class BaboonCodecException(override val message: String, cause: Throwable? = null) : Exception(message, cause) {
    class EncoderFailure(message: String, cause: Throwable? = null) : BaboonCodecException(message, cause)
    class DecoderFailure(message: String, cause: Throwable? = null) : BaboonCodecException(message, cause)
    class ConverterFailure(message: String, cause: Throwable? = null) : BaboonCodecException(message, cause)
    class CodecNotFound(message: String) : BaboonCodecException(message)
    class ConversionNotFound(message: String) : BaboonCodecException(message)
}
